### PR TITLE
Hadoop drop table to return false if table does not exist

### LINKED ISSUE #3051: Hadoop Catalog DropTable always returns true and doesn't check that the folder contains Iceberg metadata
Currently we define our drop table method here

https://github.com/apache/iceberg/blob/237b72daa211f0404b72d4cd142b4a49be33a1f0/api/src/main/java/org/apache/iceberg/catalog/Catalog.java#L277-L285

Saying that we need to return "false" if a table does not exist but the underlying delete method in hadoop only returns true unless an exception occurs when attempting to delete the underlying file. 

https://github.com/apache/iceberg/blob/237b72daa211f0404b72d4cd142b4a49be33a1f0/core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java#L268-L269

Most hadoop FS implementations I checked swallow all FileNotFound and associated exceptions and just return false.

https://github.com/apache/hadoop/blob/50dda774f13a83547e41db4e835d7106ae66c1ff/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/ChecksumFileSystem.java#L723-L728

We should probably be returning the output of fs.delete like we do in dropNamespace 
https://github.com/apache/iceberg/blob/237b72daa211f0404b72d4cd142b4a49be33a1f0/core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java#L347

I think additionally we should probably do a check that the HadoopTableOperations can actually find a metadata.json file before considering the table an Iceberg table. As is, the delete command would return true even if there is no iceberg table at that location but the location exists.

Perhaps we should be doing 

```java
// lastMetadata is only null if no metadata.json exists in this directory specified by the table
return lastMetadata != null && fs.delete(tablePath, true /* recursive */);
```

### PR DESCRIPTION
**Pull request by itachi-sharingan**
closes #3051

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java [hunk 1]
```diff
@@ -247,26 +247,25 @@ protected String defaultWarehouseLocation(TableIdentifier tableIdentifier) {
   @Override
   public boolean dropTable(TableIdentifier identifier, boolean purge) {
     if (!isValidIdentifier(identifier)) {
-      throw new NoSuchTableException("Invalid identifier: %s", identifier);
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java:250]: This is actually a dead code block for this class. If you check isValidIdentifier it is ```return true```. I think we should just leave this bit as is for now and maybe clean it up in a future pr.

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java [hunk 2]
```diff
@@ -247,26 +247,25 @@ protected String defaultWarehouseLocation(TableIdentifier tableIdentifier) {
...
     TableOperations ops = newTableOps(identifier);
-    TableMetadata lastMetadata;
-    if (purge && ops.current() != null) {
-      lastMetadata = ops.current();
-    } else {
-      lastMetadata = null;
+    TableMetadata lastMetadata = ops.current();
+    if (lastMetadata == null) {
+      LOG.error("Not an iceberg table: %s", identifier);
+      return false;
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java:259]: I would prefer keeping the returns in the same if block as the other return statement. I tend to always lean towards single or very closely located exits from functions so it's clear that you either return this or that.

Like if you had it as
```
if (lastMetadata == null) {
  LOG.error(....)
  return false
} else {
  return fs.delete(...)
}
```

**Review by RussellSpitzer — CHANGES_REQUESTED** [review]
Looks good to me, a few minor comments but we definitely need some tests to make sure we can hit all of the different return points in this code. I would suggest adding some to TestHadoopCatalog.java

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java (hunk 1 above)
Comment (itachi-sharingan) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java]: sure, have rolled back my changes for this section.

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java (hunk 2 above)
Comment (itachi-sharingan) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java]: Have tried doing something similar in the latest commit, please review.

**Comment by itachi-sharingan** [comment]
Have added test for the other case of exit because of non iceberg table, please review.

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java [hunk 3]
```diff
@@ -252,21 +252,19 @@ public boolean dropTable(TableIdentifier identifier, boolean purge) {
...
-    }
-
+    TableMetadata lastMetadata = ops.current();
     try {
-      if (purge && lastMetadata != null) {
-        // Since the data files and the metadata files may store in different locations,
-        // so it has to call dropTableData to force delete the data file.
-        CatalogUtil.dropTableData(ops.io(), lastMetadata);
+      if (lastMetadata == null) {
+        LOG.error("Not an iceberg table: %s", identifier);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java:258]: This is not an error. The API states that if the table doesn't exist, then the catalog should return false. If you want to add a debug log, that's fine. But this should not be info or error level.

File: core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java [hunk 4]
```diff
@@ -237,6 +239,24 @@ public void testDropTable() throws Exception {
...
+    String metaLocation = catalog.defaultWarehouseLocation(testTable);
+
+    FileSystem fs = Util.getFs(new Path(metaLocation), conf);
+    File tablePath = new File(warehousePath + "/db/ns1/ns2/tbl");
+    tablePath.mkdirs();
+    Assert.assertTrue(fs.isDirectory(new Path(metaLocation)));
+
+    Assert.assertFalse(catalog.dropTable(testTable));
+    Assert.assertTrue(fs.isDirectory(new Path(metaLocation)));
+    FileUtils.deleteDirectory(new File(warehousePath + "/db"));
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java:257]: No need to delete this. The temp folder will handle it.

File: core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java [hunk 5]
```diff
@@ -237,6 +239,24 @@ public void testDropTable() throws Exception {
...
+  public void testDropNonIcebergTable() throws Exception {
+    Configuration conf = new Configuration();
+    String warehousePath = temp.newFolder().getAbsolutePath();
+    HadoopCatalog catalog = new HadoopCatalog(conf, warehousePath);
+    TableIdentifier testTable = TableIdentifier.of("db", "ns1", "ns2", "tbl");
+    String metaLocation = catalog.defaultWarehouseLocation(testTable);
+
+    FileSystem fs = Util.getFs(new Path(metaLocation), conf);
+    File tablePath = new File(warehousePath + "/db/ns1/ns2/tbl");
+    tablePath.mkdirs();
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java:252]: This shouldn't mix Hadoop FileSystem calls with the Java File API. Please only use the Hadoop calls based on metadataLocation. That way you don't need to assume a location for the table.

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java (hunk 3 above)
Comment (itachi-sharingan) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java]: The reason I thought of putting it in error was calling iceberg api drop on a table which is not an iceberg table looked like an error but wasn't sure of this. But now I agree with you, so have made the log as debug now.

File: core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java (hunk 4 above)
Comment (itachi-sharingan) [reply core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java]: ok

File: core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java (hunk 5 above)
Comment (itachi-sharingan) [reply core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCatalog.java]: Makes sense, have made the change.

**Review by rdblue — APPROVED** [review]

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java (hunk 2 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java]: Looks like this is fixed to me.

**Review by RussellSpitzer — APPROVED** [review]
I would suggest adding one more test to make sure we return false when the table doesn't exist at all (no folder or anything)

**Comment by rdblue** [comment]
@itachi-sharingan, could you add the test that Russell suggested and fix the failing CI checks? Otherwise, I think this is about ready to go.

**Comment by itachi-sharingan** [comment]
@rdblue have fixed the failing CI.
@RussellSpitzer have added the check for non existent directory drop returning false in the same test.
Pls check. 

Moreover, can someone please suggest what would be a good issue to pick up next?

**Comment by rdblue** [comment]
Thanks, @itachi-sharingan! I reopened this so that I could merge it. There's no need to close a PR when it is finished.

If you're looking for another contribution, #2488 is a good one! Thanks for contributing!

**Pull request merged**